worlds/tww/randomizers/RequiredBosses.py

```diff
@@ -0,0 +1,121 @@
+        self.required_bosses: list[str] = []
+        self.banned_locations: set[str] = set()
+        self.banned_dungeons: list[str] = []
+        self.banned_bosses: list[str] = []
+
+    def validate_boss_options(self, options: TWWOptions) -> None:
+        """
+        Validate the user-defined boss options to ensure logical consistency.
+
+        :param options: The game options set by the user.
+        :raises OptionError: If the boss options are inconsistent.
+        """
+        if not options.progression_dungeons:
+            raise OptionError("You cannot make bosses required when progression dungeons are disabled.")
+
+        if len(options.included_dungeons.value & options.excluded_dungeons.value) != 0:
+            raise OptionError(
+                "A conflict was found in the lists of required and banned dungeons for required bosses mode."
+            )
+
+    def randomize_required_bosses(self) -> None:
+        """
+        Randomize the required bosses based on user-defined constraints and options.
+
+        :raises OptionError: If the randomization fails to meet user-defined constraints.
+        """
+        options = self.world.options
+
+        # Validate constraints on required bosses options.
+        self.validate_boss_options(options)
+
+        # If the user enforces a dungeon location to be priority, consider that when selecting required bosses.
+        dungeon_names = set(DUNGEON_NAMES)
+        required_dungeons = options.included_dungeons.value
+        for location_name in options.priority_locations.value:
+            dungeon_name, _ = split_location_name_by_zone(location_name)
+            if dungeon_name in dungeon_names:
+                required_dungeons.add(dungeon_name)
+
+        # Ensure we aren't prioritizing more dungeon locations than the requested number of required bosses.
+        num_required_bosses = options.num_required_bosses
+        if len(required_dungeons) > num_required_bosses:
+            raise OptionError(
+                "Could not select required bosses to satisfy options set by the user. "
+                "There are more dungeons with priority locations than the desired number of required bosses."
+            )
+
+        # Ensure that after removing excluded dungeons, we still have enough to satisfy user options.
+        num_remaining = num_required_bosses - len(required_dungeons)
+        remaining_dungeon_options = dungeon_names - required_dungeons - options.excluded_dungeons.value
+        if len(remaining_dungeon_options) < num_remaining:
+            raise OptionError(
+                "Could not select required bosses to satisfy options set by the user. "
+                "After removing the excluded dungeons, there are not enough to meet the desired number of required"
+                "bosses."
+            )
+
+        # Finish selecting required bosses.
+        required_dungeons.update(self.world.random.sample(list(remaining_dungeon_options), num_remaining))
```

Comment:
`remaining_dungeon_options` is a `set[str]`, so this was picking required bosses nondeterministically.
```suggestion
        required_dungeons.update(self.world.random.sample(sorted(remaining_dungeon_options), num_remaining))
```